### DISCUSSION TITLE
calling metadata.drop_all() drops tables on the database but does not remove the tables from the metadata object

### DISCUSSION THREAD

**Discussion by pscholle** [sqlalchemy/sqlalchemy#7293] [Usage Questions] [answered]
Hello everyone!

Just looking for some clarity on the correct use of the metadata object.

Whats getting me down right now is the drop_all function.
The tables are dropped on the database side as expected, however the table objects are still stored in the metadata object. Calling reflect() after drop all does not update the metadata object (I assume it will only add tables and not remove any that are no longer present on the database?)

I expect that when drop_all is called, that there are also no more tables stored in the metadata, but this is not the case with me. 

I am using mssql driver.

```python
def test_drop_all():
    m = MetaData()
    # get any tables already in the database
    m.reflect()

    print(m.tables.keys())  # should be empty

    # add tables to metadata
    Table('test', m,
          Column('id', Integer, primary_key=True),
          Column("data", VARBINARY, nullable=False))

    m.create_all()

    print(m.tables.keys())  # ['test']

    m.drop_all()  # the database is empty after this call but,
    print(m.tables.keys())  # still has ['test']
```

Currently I am creating a new metadata object and calling reflect after drop_all (in case new tables were created by someone else). Is this the correct usage?

Ideally I don't want to be creating new metadata objects all the time but rather work with a single instance. Actually I am declaring the metadata object as a global variable and initializing it once then using it for deleting and creating tables etc.

Cheers and looking forward to any help!
P

**Comment by CaselIT** [comment, marked as answer]
Hi,

That is the expected behaviour. You can use clear to remove the tables from the metadata https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=metadata#sqlalchemy.schema.MetaData.clear
